From d3f16ef8fb915b30a459c1c191f6d0c63b38c491 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Sat, 21 Jun 2025 22:35:15 +0100
Subject: [PATCH 08/11] opensslsetup: add support for portable libcrypto

Sometimes host OS does not have OpenSSL or it is not FIPS capable. Add
support to find a co-located copy of OpenSSl and load that by default.

The intention is to add support for running CNI binaries on host with
just glibc 2.34 available.
---
 .../internal/opensslsetup/opensslsetup.go     | 59 +++++++++++++++++++
 1 file changed, 59 insertions(+)

diff --git a/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go b/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go
index 14b217a6e2..2d917b8f68 100644
--- a/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go
+++ b/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go
@@ -10,11 +10,20 @@
 package opensslsetup
 
 import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
 	"syscall"
 
 	"github.com/golang-fips/openssl/v2/osslsetup"
 )
 
+// #cgo LDFLAGS: -lc
+// #include <gnu/libc-version.h>
+import "C"
+
 // knownVersions is a list of supported and well-known libcrypto.so suffixes in decreasing version order.
 // FreeBSD library version numbering does not directly align to the version of osslsetup.
 // Its preferred search order is 11 -> 111.
@@ -29,11 +38,61 @@ func init() {
 	}
 }
 
+// Check for co-located libcrypto
+func portablecrypto() (string, error) {
+	executable, err := os.Executable()
+	if err != nil {
+		return "", err
+	}
+	portabledir := filepath.Join(filepath.Dir(executable), ".cgr")
+	portableconf := filepath.Join(portabledir, "openssl.cnf")
+
+	// Check for co-located openssl.cnf as an indicator,
+	// should also contain everything else too ca.cnf, fipsmodule.cnf, libcrypto.so.3, fips.so, etc
+	if _, err := os.Stat(portableconf); err != nil {
+		return "", err
+	}
+
+	// Check compatible glibc version
+	// Currently our jitterentropy source requires glibc 2.34, and thus also our libcrypto.so and fips.so
+	version := strings.Split(C.GoString(C.gnu_get_libc_version()), ".")
+	major, err := strconv.Atoi(version[0])
+	if err != nil {
+		return "", err
+	}
+	minor, err := strconv.Atoi(version[1])
+	if err != nil {
+		return "", err
+	}
+	if major != 2 {
+		return "", errors.New("glibc major version is not 2")
+	}
+	if minor < 34 {
+		return "", errors.New("glibc minor version less than 34")
+	}
+
+	// Setup portable libcrypto
+	if err = os.Setenv("OPENSSL_CONF", portableconf); err != nil {
+		return "", err
+	}
+	if os.Setenv("OPENSSL_CONF_INCLUDE", portabledir) != nil {
+		return "", err
+	}
+	if os.Setenv("OPENSSL_MODULES", portabledir) != nil {
+		return "", err
+	}
+
+	return filepath.Join(portabledir, "libcrypto.so.3"), nil
+}
+
 // library returns the name of the OpenSSL library to use.
 // It first checks the environment variable GO_OPENSSL_VERSION_OVERRIDE.
 // If that is not set, it searches a well-known list of library names.
 // If no library is found, it returns "libcrypto.so".
 func library() string {
+	if portable, err := portablecrypto(); err == nil {
+		return portable
+	}
 	if version, _ := syscall.Getenv("GO_OPENSSL_VERSION_OVERRIDE"); version != "" {
 		return lcryptoPrefix + version
 	}
-- 
2.51.0

